modules/javafx.web/src/main/java/com/sun/javafx/webkit/prism/WCGraphicsPrismContext.java

```diff
@@ -523,6 +528,52 @@ public void fillRect(final float x, final float y, final float w, final float h,
                     final NGRectangle node = new NGRectangle();
                     node.updateRectangle(x, y, w, h, 0, 0);
                     render(g, shadow, paint, null, node);
+                } else if(state.getClipMaskImageNoClone() != null) {
+                    Rectangle rect = new Rectangle((int) x, (int) y, (int) w, (int) h);
+                    Rectangle transformedRect = new Rectangle();
+                    state.getTransformNoClone().transform(rect, transformedRect);
+                    RTTexture paintRtTexture = g.getResourceFactory().createRTTexture(
+                            (int) Math.ceil(transformedRect.width),
+                            (int) Math.ceil(transformedRect.height),
+                            Texture.WrapMode.CLAMP_NOT_NEEDED);
+                    Graphics g1 = paintRtTexture.createGraphics();
+                    g1.setPaint(paint);
+                    g1.setTransform(PrismGraphicsManager.getPixelScaleTransform());
+                    g1.fillRect(0, 0, paintRtTexture.getContentWidth(), paintRtTexture.getContentHeight());
+
+                    RTImage maskImage = (RTImage) (state.getClipMaskImageNoClone());
+                    Image nativeMaskImage = Image.fromByteBgraPreData(maskImage.getPixelBuffer(), maskImage.getWidth(),
+                            maskImage.getHeight(), 0, maskImage.getPixelScale());
+                    Texture maskTexture = g.getResourceFactory().createTexture(nativeMaskImage, Texture.Usage.STATIC,
+                            Texture.WrapMode.CLAMP_NOT_NEEDED);
+                    RTTexture maskRtTexture = g.getResourceFactory().createRTTexture(nativeMaskImage.getWidth(),
```

Comment:
Why do you need to create a second RTT here? I would have thought you could use the mask texture directly (you may need to scale the image in order to do that).

Reply:
main problem with not using second RTTexture is the interface MaskTextureGraphics, which accept RTTexture only. also it difficult to add new API taking Texture instead of RTTexture given that SW Pipeline and J2D pipeline has direct dependency on RTTexture interface.